proposals/nnnn-rpc-design-refactoring.md

```diff
@@ -0,0 +1,147 @@
+# RPC design refactoring
+
+* Proposal: [SDL-NNNN](nnnn-rpc-design-refactoring.md)
+* Author: [Alexander Kutsan](https://github.com/LuxoftAKutsan)
+* Status: **Awaiting review**
+* Impacted Platforms: [Core]
+
+## Introduction
+
+Rework design of application manager layer to:
+ - Reduce coupling between components
+ - Reduce responsibility of ApplicationManager class
+ - Provide plugin architecture for easy adding additional functionality and RPC groups   
+
+## Motivation
+
+SDL core project is actively implements new features and new functionality. 
+To keep project on track and reduce rick of regression during implementation
+some refactoring of application manager layer should be done. 
+
+### Reduce coupling between components and use dependency injection practice
+
+Now Application manager is service locator and need to passed in each component
+even if component does not need application manager, but need some other classes that owned by application manager.
+During construction components should explicitly receive all dependencies. 
+It will prevent unwanted access and improve encapsulation. 
+
+
+### Reduce responsibility of ApplicationManager class 
+Not application manager class have to many responsibilities:
+ - Service location
+ - Mobile/HMI RPC processing 
+ - Mobile/HMI RPC sending 
+ - Registration and managing mobile applications
+ - Handle different states of SDL (low voltage/ video streaming / audio_path_throw ...) 
+ 
+ Purpose of this refactoring is to extract handling/sending RPC's to separate module. 
+ Many SDL commands require application manager only for sending RPC's and do not need to 
+ know anything related to other application manager responsibility. 
+ Also extracting RPC processing to separate module will reduce regression risk of changing any functionality that now in ApplicationManager.
+ 
+### Provide plugin architecture for easy adding additional functionality and RPC groups   
+
+SDL RPC's can be splitted to some groups of RPC's related so certain functionality (for example RC or video streaming).
+Usually this RPC's works with some scope functionality that not required for other RPC's so should not be exposed. 
```

Comment:
**_Usually these RPC's work with some functionality that is not required_** for other RPC's so should not be exposed

Reply:
fixed